airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py

```diff
@@ -401,3 +435,124 @@ def upload_metadata_to_gcs(bucket_name: str, metadata_file_path: Path, validator
+
+    version_path = get_metadata_remote_file_path(docker_repository, connector_version)
+    rc_path = get_metadata_remote_file_path(docker_repository, RELEASE_CANDIDATE_GCS_FOLDER_NAME)
+
+    version_blob = bucket.blob(version_path)
+    rc_blob = bucket.blob(rc_path)
+
+    if not version_blob.exists():
+        raise FileNotFoundError(f"Version metadata file {version_path} does not exist in the bucket. ")
+    if not rc_blob.exists():
+        raise FileNotFoundError(f"Release candidate metadata file {rc_path} does not exist in the bucket. ")
+    if rc_blob.md5_hash != version_blob.md5_hash:
+        raise ValueError(
+            f"Release candidate metadata file {rc_path} hash does not match the version metadata file {version_path} hash. Unsafe to delete."
+        )
+
+    deleted_files = []
+    rc_blob.delete()
+    deleted_files.append(
+        DeletedFile(
+            id="release_candidate_metadata",
+            deleted=True,
+            description="release candidate metadata",
+            blob_id=rc_blob.id,
+        )
+    )
+    version_blob.delete()
+    deleted_files.append(
+        DeletedFile(
+            id="version_metadata",
+            deleted=True,
+            description="versioned metadata",
+            blob_id=version_blob.id,
+        )
+    )
+
+    return MetadataDeleteInfo(
+        metadata_deleted=True,
+        deleted_files=deleted_files,
+    )
+
+
+def promote_release_candidate_in_gcs(
+    bucket_name: str, docker_repository, connector_version
+) -> Tuple[MetadataUploadInfo, MetadataDeleteInfo]:
+    """Promote a release candidate to the latest version in a GCS bucket.
+    The release candidate metadata file will be copied to the latest metadata file and then deleted.
+    We first check that the release candidate metadata file hash matches the version metadata file hash.
+
+    Args:
+        bucket_name (str): Name of the GCS bucket from which the release candidate will be deleted.
+        docker_repository (str): Name of the connector docker image.
+        connector_version (str): Version of the connector.
+    """
+
+    storage_client = _get_storage_client()
+    bucket = storage_client.bucket(bucket_name)
+
+    version_path = get_metadata_remote_file_path(docker_repository, connector_version)
+    rc_path = get_metadata_remote_file_path(docker_repository, RELEASE_CANDIDATE_GCS_FOLDER_NAME)
+    latest_path = get_metadata_remote_file_path(docker_repository, LATEST_GCS_FOLDER_NAME)
+
+    version_blob = bucket.blob(version_path)
+    latest_blob = bucket.blob(latest_path)
+    rc_blob = bucket.blob(rc_path)
+
+    if not version_blob.exists():
+        raise FileNotFoundError(f"Version metadata file {version_path} does not exist in the bucket.")
+    if not rc_blob.exists():
+        raise FileNotFoundError(f"Release candidate metadata file {rc_path} does not exist in the bucket.")
+
+    if rc_blob.md5_hash != version_blob.md5_hash:
+        raise ValueError(
```

Comment:
To check my understanding - when a connector change is merged, the blob is pushed to both the RC directory and a directory for that version. So if we hit this code path it means that somehow there was a re-upload to the RC directory without a corresponding one to the version directory? If that's the case sounds like we really don't expect this to happen, but I'm curious what you think the solution would be if it did. Would be great to add that to the error message.

It would be nice to use pointers instead of uploading to multiple locations. I expect there's a good reason we're not doing that?

Reply:
> If that's the case sounds like we really don't expect this to happen.

Yes this is not something which is expected to happen.
Inconsistent MD5 might happen on incomplete publish flow.
I think this check is important as it makes sure we're promoting/rolling back the right version.

> It would be nice to use pointers instead of uploading to multiple locations. I expect there's a good reason we're not doing that?

Yes pointers would be better, but our whole registry generation logic is based on sensors triggered by upload events to GCS to specific path. It's more performant to identify existing RCs by listing blobs with a specific `**/release_candidate/metadata.yaml` path than continually scanning all our metadata and regenerating a list of RC version metadata path. I'm open to change this approach later if we find it too fragile.